doc/common_pitfalls.rst

```diff
@@ -0,0 +1,232 @@
+    GradientBoostingClassifier(random_state=1)
+    >>> X_test_selected = select.transform(X_test)
+    >>> y_pred = gbc.predict(X_test_selected)
+    >>> score = accuracy_score(y_test, y_pred)
+    >>> print(f"Accuracy: {score}")
+    Accuracy: 0.46
+
+Another way to prevent data leakage is to use the
+:class:`~sklearn.pipeline.Pipeline` to chain together the feature selection
+and model estimators. This pipeline can then be fed into a cross validation
+function such as :func:`~sklearn.model_selection.cross_val_score`. The
+pipeline ensures that only the training data when performing `fit` and
+the test data will only be used for calculating the accuracy score.
+Using :func:`~sklearn.model_selection.cross_val_score` also provides us
+information about the variance of the accuracy score::
+
+    >>> from sklearn.pipeline import make_pipeline
+    >>> from sklearn.model_selection import cross_val_score
+    >>> pipeline = make_pipeline(SelectKBest(k=25),
+    ...                          GradientBoostingClassifier(random_state=1))
+    >>> scores = cross_val_score(pipeline, X, y)
+    >>> print(f"Mean Accuracy: {scores.mean():.2f}+/-{scores.std():.2f}")
+    Mean Accuracy: 0.45+/-0.07
+
+Data leakage during imputation
+------------------------------
+
+There are a number of methods to impute missing values in data. For example,
+:class:`~sklearn.impute.SimpleImputer` allows you to replace missing values
+with the mean of each feature. Only the train data should be used to
+calculate this mean value, as including the test data in the mean calculation
+will introduce information about the test data into the model.
+
+To demonstrate this, we will use the :ref:`diabetes_dataset` and
+artificially introduce 15% missing values::
+
+    >>> import numpy as np
+    >>> from sklearn.datasets import load_diabetes
+    >>> X, y = load_diabetes(return_X_y=True)
+    >>> n_samples, n_features = X.shape
+    >>> rng = np.random.RandomState(42)
+    >>> missing_mask = rng.binomial(n=1, p=0.2, size=(n_samples, n_features))
+    >>> missing_mask = missing_mask.astype(bool)
+    >>> X_missing = X.copy()
+    >>> X_missing[missing_mask] = np.nan
+
+**Wrong**
+
+Using all the data to calculate the feature means, to replace the missing
+values with, results in an overly optimsitic :math:`R^2`::
+
+    >>> from sklearn.impute import KNNImputer
+    >>> from sklearn.model_selection import train_test_split
+    >>> from sklearn.ensemble import GradientBoostingRegressor
+    >>> from sklearn.metrics import r2_score
+    >>> X_impute = KNNImputer().fit_transform(X_missing)
+    >>> X_train, X_test, y_train, y_test = train_test_split(
+    ...     X_impute, y, random_state=42)
+    >>> gbr = GradientBoostingRegressor(random_state=1)
+    >>> gbr.fit(X_train, y_train)
+    GradientBoostingRegressor(random_state=1)
+    >>> y_pred = gbr.predict(X_test)
+    >>> score = r2_score(y_test, y_pred)
+    >>> print(f"R2 score: {score:.3f}")
+    R2 score: 0.332
```

Comment:
this is wrong but it's not revealed by the number here. As suggested (by I don't remember who) you should not have features missing fully at random (MAR) but rather MNAR (not at random). Typically you should add missing values more for certain values of y (eg in a given class). Did you try?

Reply:
It's mentioned here: https://github.com/scikit-learn/scikit-learn/pull/18426#discussion_r492765211

What do you mean by wrong? From @NicolasHug's comment I understand that MNAR is more realistic. The KNN imputation paper referenced in the user guide ( Missing value estimation methods for DNA microarrays) evaluated the method by deleteing 1-20% of the data at random, so I don't think KNN method assumes MNAR?

Nicholas tried above but says it wasn't super compelling. I'll try MNAR and see how much difference it makes.

Reply:
I mean wrong to reference the section title "**wrong**" 

try something brutal. Remove values for one features if and only if y = 1 for binary classification problem. You should see then a clear effect.

Reply:
So I tried with the regression diabetes dataset, only putting missing values in the smallest 100 `y` samples and either no missing values in the rest or much less (e.g., 10x less likely to be missing). There wasn't a clear effect.
I also tried with the classification breast cancer dataset, only putting missing values for `y==1` and again, no clear effect.

Maybe I misunderstood, I put MNAR in all the data and then split. Should I be putting missing values in only in the train data subset (this doesn't sound right though)?

Reply:
> Maybe I misunderstood, I put MNAR in all the data and then split. 

yes that's what I had in mind.

try this

```
import numpy as np
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.metrics import accuracy_score

n_samples, n_features, n_classes = 1000, 5, 2
rng = np.random.RandomState(42)
X = rng.standard_normal((n_samples, n_features))
y = rng.choice(n_classes, n_samples)

# **wrong**

missing_mask = rng.binomial(n=1, p=0.2, size=(n_samples, n_features))
missing_mask = missing_mask.astype(bool)

missing_mask[y == 0, 2] = True  # MNAR

X_missing = X.copy()
X_missing[missing_mask] = np.nan

X_impute = SimpleImputer().fit_transform(X_missing)
X_train, X_test, y_train, y_test = train_test_split(
    X_impute, y, random_state=42)
clf = GradientBoostingClassifier(random_state=1)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
score = accuracy_score(y_test, y_pred)
print(f"Accuracy score: {score:.3f}")

# **right**

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
clf = make_pipeline(SimpleImputer(), GradientBoostingClassifier(random_state=1))
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
score = accuracy_score(y_test, y_pred)
print(f"Accuracy score: {score:.3f}")
```

Reply:
Ah thanks, I was too ambitious trying to use real data...

Reply:
Actually, there isn't a big difference in accuracy scores (see latest commit)

fyi:
```
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
```
should be:
```
X_train, X_test, y_train, y_test = train_test_split(X_missing, y, random_state=42)
```

Reply:
grrrr indeed the effect is not so clear... i don't know what else to suggest except just moving on and deleting the paragraph on imputation...

Reply:
I really repeat the information in the 2 sections currently (the feature selection and imputation examples) so I think deleting the imputation example is fine.